### PR TITLE
Feature/160451496/user login jwt

### PR DESCRIPTION
**Pull request by udochukwu**
**What does this PR do?**
This PR logs In a user into Authors Haven

Description of Task to be completed?

- Create a method for user login

- Validate user login parameters.

- Write tests for user login

**What are the relevant pivotal tracker stories?**
https://www.pivotaltracker.com/story/show/160451496

**Screenshots (if appropriate)**
N/A
**Questions:```**
N/A
N/A

### CODE REVIEW

File: server/middlewares/userValidations.js [hunk 1]
```diff
@@ -16,9 +16,15 @@ class UserValidations {
    */
   static validateSignup(req, res, next) {
     // this function validates the input the user sends when trying to register
-    // this regex removes all the white space characters in the email and username
-    req.body.email = req.body.email.toLowerCase().replace(/\s/g, '');
-    req.body.username = req.body.username.replace(/\s/g, '');
+
+    // check if email and username is defined
```
Comment (udochukwu) [inline server/middlewares/userValidations.js]: I added this block of code to check if username and email are defined before converting them to lowercase.

File: server/middlewares/userValidations.js [hunk 2]
```diff
@@ -16,9 +16,15 @@ class UserValidations {
...
+    if (req.body.email && req.body.username) {
+      /* this regex removes all the white space characters in the email and username, and also
+       converts  to lowercase */
+      req.body.email = req.body.email.toLowerCase().replace(/\s/g, '');
+      req.body.username = req.body.username.replace(/\s/g, '');
+    }
+
```
Comment (udochukwu) [inline server/middlewares/userValidations.js:27]: I added this block of code to check if username and email are defined before converting them to lowercase.

**Comment by kampkelly** [comment]
LGTM

**Review by slimsolz — COMMENTED** [review]
LGTM

**Review by onwuvic — COMMENTED** [review]
LGTM

File: server/controllers/userController.js [hunk 3]
```diff
@@ -29,4 +31,39 @@ export default class UserController {
...
+      .then((user) => {
+        if (!user) {
+          return res.status(404)
+            .send({ success: false, message: 'Invalid Email/Username or password' });
+        }
+        const { id, username } = user;
+        if (bcrypt.compareSync(req.body.password, user.password)) {
+          const token = generateToken({ id, username });
+          res.json({
+            success: true, message: 'successfully logged in!', user, token
```
Comment (s-akinrele) [inline server/controllers/userController.js]: @udochukwu updated the text to capitalised case `Successfully`

File: server/middlewares/userValidations.js [hunk 4]
```diff
@@ -16,9 +16,15 @@ class UserValidations {
...
     // this function validates the input the user sends when trying to register
-    // this regex removes all the white space characters in the email and username
-    req.body.email = req.body.email.toLowerCase().replace(/\s/g, '');
-    req.body.username = req.body.username.replace(/\s/g, '');
+
+    // check if email and username is defined
+    if (req.body.email && req.body.username) {
+      /* this regex removes all the white space characters in the email and username, and also
+       converts  to lowercase */
+      req.body.email = req.body.email.toLowerCase().replace(/\s/g, '');
```
Comment (s-akinrele) [inline server/middlewares/userValidations.js:24]: ```req.body.email &&  req.body.email.toLowerCase().replace(/\s/g, '');```

File: server/middlewares/userValidations.js (hunk 2 above)
Comment (s-akinrele) [reply server/middlewares/userValidations.js:27]: good catch @udochukwu

File: server/tests/user.test.js [hunk 5]
```diff
@@ -207,4 +207,41 @@ describe('Users', () => {
...
+        .post('/api/v1/login')
+        .send({ username: 'testuser', password: 'password' })
+        .end((err, res) => {
+          expect(res.body).to.be.an('object');
+          assert.isNotNull(res.body.token);
+          done();
+        });
+    });
+
+    it('Should return error message when user tries to login in with invalid email/username', (done) => {
```
Comment (s-akinrele) [inline server/tests/user.test.js:223]: @udochukwu please use expect, instead since it provides the same functionality

**Review by s-akinrele — APPROVED** [review]

**Pull request merged**